### ISSUE TITLE
Error with fused function dispatch on Linux with Python 2.7 and NumPy 1.16.3

### ISSUE THREAD

**Issue by mobiusklein** [numpy/numpy#13430]
This problem is related to Cython, but given that it interacts with the initialization of NumPy's infrastructure, it might make sense to report here too (https://github.com/cython/cython/issues/2932).

### Reproducing code example:

Minimal Reproducible Example Requires Compilation: https://gist.github.com/mobiusklein/808d101ea228a71af02e9b5f4a8a0caa

It was observed that using Python 2.7 on Linux with NumPy 1.16, invoking an unspecialized fused function where `np.ndarray` is in the fused type produces `RuntimeError: implement_array_function method already has a docstring`.

The gist includes the source code for building `numfoo`, and a short test script to demonstrate the behavior:

```python
import numfoo
import numpy as np

data = [1.2, 2.4]

# The data are a plain list
print(data)

# Invoking the fused function explicitly with a list works
print(numfoo.make_array[list](data))

# convert it into a NumPy array
data = np.array(data)

# Invoking the fused function explicitly with an np.ndarray works
print(numfoo.make_array[np.ndarray](data))

# This fails
print(numfoo.make_array(data))
```
This does not happen with an earlier NumPy version, it does not happen with a newer version of Python, and it does not appear to happen on Windows. That it blows up inside the fused function dispatch machinery but with seemingly an initialization-like problem in NumPy makes it hard to tell where I should be reporting this bug.

Note: Compilation was does using Cython 0.29.7

### Error message:

```
/usr/lib/python2.7/site-packages/numpy/__init__.py in <module>()
    140     from . import _distributor_init
    141 
--> 142     from . import core
    143     from .core import *
    144     from . import compat

/usr/lib/python2.7/site-packages/numpy/core/__init__.py in <module>()
     38 
     39 try:
---> 40     from . import multiarray
     41 except ImportError as exc:
     42     import sys

/usr/lib/python2.7/site-packages/numpy/core/multiarray.py in <module>()
     10 import warnings
     11 
---> 12 from . import overrides
     13 from . import _multiarray_umath
     14 import numpy as np

/usr/lib/python2.7/site-packages/numpy/core/overrides.py in <module>()
     44     ------
     45     TypeError : if no implementation is found.
---> 46     """)
     47 
     48 
```
### Numpy/Python version information:

Python: 2.7.13
NumPy: 1.16.3
Platform: Linux

**Comment by charris** [comment]
Thanks for the report. The easy fix here is to wait on a Cython fix, but we could possibly revert something if that that takes too long. Please keep us updated.

**Comment by charris** [comment]
Where exactly is the fused function dispatch located? NumPy itself doesn't use it AFAIK.

**Comment by mobiusklein** [comment]
The fused function is in third-party code, stripped down to `numfoo.pyx` in the linked Gist. It's that during the fused function's dispatching process, which is entirely inside Cython's control, something causes an exception inside NumPy's internals.

It looks like it is related to dynamic creation of user-friendly looking wrappers of C functions, but the traceback shown in the issue looks like the sort of thing that ought to happen upon importing something, when both NumPy's Python layer and its C layer have both been imported already.

**Comment by seberg** [comment]
OK, I can reproduce this, but I am still not sure why there is interaction with `overrides.py` here. Should probably wait for the cython folks to give some pointers?

**Comment by JamesXNelson** [comment]
subscribing to issue.

I'm hitting this when trying to load numpy through jpy.  I'm fairly newbish in the python scene; would this be exacerbated by loading numpy twice through different mechanisms?

**Comment by JamesXNelson** [comment]
ah.  disregard my report; my error was due to trying to reload jpy, which triggers manual double-inclusion of python modules (or some other terrible JNI voodoo) that causes the issue.

**Comment by charris** [comment]
Thanks for the followup.

**Comment by mobiusklein** [comment]
To be clear, the original issue still stands due to an as yet un-understood interaction between Cython's fused types and NumPy 1.16 on Python 2.7 and Linux, and had nothing to do with jpy, so I don't think this issue should be closed.

That said, I got no response from Cython's developers on this issue when I opened it with them, and I don't see any place in the internals of Cython's fused functions that could cause it, so I've had to work around the problem by releasing a different algorithm implementation for people in the 1.16/2.7/Linux camp.

Given that this bug has five months to live anyway, I don't know if you want to keep the issue around for posterity.

**Comment by bacon-cheeseburger** [comment]
I have this problem as well. What is the fix/workaround/patch?

**Comment by mobiusklein** [comment]
There hasn't been a response from Cython's maintainers yet. If you're getting the same Platform/Python Version/NumPy version-related issue, rolling back NumPy should fix the problem. It only appears in 1.16, so reverting to 1.15 should work.

Alternatively, if its your own fused function, and you're not concerned with the small performance hit for affected users, you could use Cython's conditional compilation mechanism to generate different code for Linux Py2.7 like [this workaround](https://github.com/mobiusklein/pynumpress/blob/master/pynumpress/pynumpress.pyx#L45-L106).

**Comment by bacon-cheeseburger** [comment]
Does current numpy 1.17.0 also have this problem?

**Comment by rgommers** [comment]
Since this is about `__array_function__` which was marked experimental in numpy 1.16.x, and 1.15 and 1.17 work, I'm not sure it still makes sense to spend effort on this. If someone has and needs a fix, we can still put it in the 1.16.x branch - we'll make at least one more release from that, and people will likely be relying on it for py27 support for a while longer. If it doesn't materialize, we can close this once the 1.16.x LTS branch is no longer supported in 2020.

> Does current numpy 1.17.0 also have this problem?

It shouldn't, since it's Python 2.7 specific and NumPy 1.17.0 does not support Python 2.7 anymore.

**Comment by seberg** [comment]
I cannot reproduce this anymore with cython3.0a0 and numpy 1.16.2. I am tempted to think that cython has fixed this and will close for now (Wil lmake a note at the cython issue). If it remains, lets reopen.
